### ISSUE TITLE
Performance tests on Jenkins are failing after #6638

### ISSUE THREAD

**Issue by wojtek-t** [GoogleCloudPlatform/kubernetes#7216]
After #6638 performance tests are failing. Two example failures:

/go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/density.go:323
event:' &{{ } {my-hostname-density3000-00320825-e94d-11e4-899e-42010af015s10vx.13d77c120229a7c3  e2e-density3000-00320825-e94d-11e4-899e-42010af01555 /api/v1beta3/namespaces/e2e-density3000-00320825-e94d-11e4-899e-42010af01555/events/my-hostname-density3000-00320825-e94d-11e4-899e-42010af015s10vx.13d77c120229a7c3 1971aa23-e94e-11e4-a0c6-42010af037e1 29803 2015-04-22 17:17:14 -0700 PDT 2015-04-22 18:17:17 -0700 PDT map[] map[]} {Pod e2e-density3000-00320825-e94d-11e4-899e-42010af01555 my-hostname-density3000-00320825-e94d-11e4-899e-42010af015s10vx 06973ed9-e94d-11e4-a0c6-42010af037e1 v1beta3 14198 } failedScheduling Binding rejected: binding "my-hostname-density3000-00320825-e94d-11e4-899e-42010af015s10vx" cannot be updated: pod my-hostname-density3000-00320825-e94d-11e4-899e-42010af015s10vx is already assigned to host "e2e-scalability-minion-trrc" {scheduler } 2015-04-22 17:10:23 -0700 PDT 2015-04-22 17:10:23 -0700 PDT %!s(int=1)}', reason: 'my-hostname-density3000-00320825-e94d-11e4-899e-42010af015s10vx.13d77c120229a7c3', message: 'Binding rejected: binding "my-hostname-density3000-00320825-e94d-11e4-899e-42010af015s10vx" cannot be updated: pod my-hostname-density3000-00320825-e94d-11e4-899e-42010af015s10vx is already assigned to host "e2e-scalability-minion-trrc"', field path: ''
Expected
    <string>: failedScheduling
not to contain substring
    <string>: fail

/go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/density.go:323
event:' &{{ } {my-hostname-density3000-9c626d26-e96a-11e4-bfc2-42010af0156p9pu.13d7886baeee18d1  e2e-density3000-9c626d26-e96a-11e4-bfc2-42010af01555 /api/v1beta3/namespaces/e2e-density3000-9c626d26-e96a-11e4-bfc2-42010af01555/events/my-hostname-density3000-9c626d26-e96a-11e4-bfc2-42010af0156p9pu.13d7886baeee18d1 ce47faa3-e96c-11e4-b00c-42010af0c464 47760 2015-04-22 20:57:05 -0700 PDT 2015-04-22 21:57:05 -0700 PDT map[] map[]} {Pod e2e-density3000-9c626d26-e96a-11e4-bfc2-42010af01555 my-hostname-density3000-9c626d26-e96a-11e4-bfc2-42010af0156p9pu f280a633-e96a-11e4-b00c-42010af0c464 v1beta3 46779 spec.containers{my-hostname-density3000-9c626d26-e96a-11e4-bfc2-42010af01555}} killing Killing b90b4793db3fae957095f320ad975ca222db22d3cc9b0c1e643860bb60e5c450 {kubelet e2e-scalability-minion-d6p6} 2015-04-22 20:56:42 -0700 PDT 2015-04-22 20:56:42 -0700 PDT %!s(int=1)}', reason: 'my-hostname-density3000-9c626d26-e96a-11e4-bfc2-42010af0156p9pu.13d7886baeee18d1', message: 'Killing b90b4793db3fae957095f320ad975ca222db22d3cc9b0c1e643860bb60e5c450', field path: 'spec.containers{my-hostname-density3000-9c626d26-e96a-11e4-bfc2-42010af01555}'
Expected
    <string>: killing
not to contain substring
    <string>: kill

BTW, I'm not 100% sure if those are real failures (especially the first one seems to be false-positive).

cc @fgrzadkowski @rrati @lavalamp @timothysc

**Comment by rrati** [comment]
I'm not sure what is causing the first error but that seems like an issue to me.  The 2nd is likely indicating a pod is killed when they are starting up.  Also a bad thing.  I added event checking to the density tests that looks for error fail or kill messages while pods are starting up.

**Comment by timothysc** [comment]
Perhaps we should just log for now, and address issues as they are unearthed.

**Comment by wojtek-t** [comment]
@timothysc - SGTM - I can prepare a PR for it tomorrow (if noone picks it up before).

**Comment by bprashanth** [comment]
Some suggestions:
1. I always find the logs of the scheduler/rcmanager etc more useful to debug something like this. How about we offload all of /var/log/kube* into a gs bucket tagged by build id at the end of every run? @quinton-hoole     
2. It would be even more helpful if we could collect resource traces (do we already? I mean just something simple like running atop) Looking through some failures:
```
{replication-controller } 2015-04-23 02:42:28 -0700 PDT 2015-04-23 02:42:28 -0700 PDT %!s(int=1)}', reason: 'my-hostname-density3000-e9b84f3a-e99c-11e4-9e76-42010af01555.13d79b49f4be3083', message: 'Error creating: The POST operation against Pod could not be completed at this time, please try again.', 
```
and 
```
    Get https://104.197.17.132/api/v1beta3/namespaces/e2e-density3000-3af9afcd-e960-11e4-a9bf-42010af01555/pods?labelSelector=name%3Dmy-hostname-density3000-3af9afcd-e960-11e4-a9bf-42010af01555: dial tcp 104.197.17.132:443: connection refused
```
look like something server side got unresponsive, could be because of cpu or ram.    
3. The density test uses a DeleteRC function that is essentially a stop rc (in the kubectl sense). Maybe it can just use the kubectl libraries (https://github.com/GoogleCloudPlatform/kubernetes/blob/master/pkg/kubectl/stop.go#L66)? Particularly, this update can fail:
https://github.com/GoogleCloudPlatform/kubernetes/blob/master/test/e2e/density.go#L63

And that would be normal, in fact I think that might've been what happened here:
```
Expected
    <int>: 0
to equal
    <int>: 300
```    
That error looks like a CAS failure    
4. I also see:
```
Expected
    <string>: failedCreate
not to contain substring
    <string>: fail
```
from: https://github.com/GoogleCloudPlatform/kubernetes/blob/master/pkg/controller/replication_controller.go#L98

I would think it's (arguably) ok to have a failed pod create that subsequently succeeded since that's what the rcs are meant for, maybe checking events for the failed substring is too draconian?

**Comment by rrati** [comment]
I'm not in favor of changing the error messages to be log messages.  We should be aiming for the tests to find issues and regressions.  If the test is failing then that is an indication it is doing its job and we need to fix the core issue.

@bprashanth DeleteRC exists because using the kubectl stop method wasn't (still isn't?) an api call and using kubectl functions wasn't feasible at the time.  With delete changing to be stop by default, DeleteRC may not be needed anymore.  DeleteRC shouldn't be needed, but at the time it was written a minor re-architecture was needed to expose stop like functionality.

I don't think it's ok to have a pod fail to be created or an api timeout in this suite.  Those are indications of a possibly over-stressed system, like the apiserver getting overloaded.  This test is designed to show that the system can work under heavy load, so a pod failing to be created or an api timeout is a failure that should be investigated and fixed imo.

I do realize things get fuzzy with rcs re-creating pods, but one of the reasons we wanted to add event checking was because of a condition where pods were having issues running and were being killed and restarted.  From the kube api perspective the pods were marked as running, but underneath there were a host of failures that were clear in the event stream.

It's also possible that the event checking isn't looking for the right strings, and if that is the case I would rather we refine what we are looking for.

**Comment by lavalamp** [comment]
@rrati:

> Binding rejected: binding "my-hostname-density3000-00320825-e94d-11e4-899e-42010af015s10vx" cannot be updated: pod my-hostname-density3000-00320825-e94d-11e4-899e-42010af015s10vx is already assigned to host "e2e-scalability-minion-trrc"', field path: ''
Expected
> : failedScheduling
> not to contain substring
> : fail

This is clearly a test error and not a system error.

**Comment by rrati** [comment]
@lavalamp Can you explain why?  Is this is acceptable, why are we generating an event saying there is a failure?

**Comment by bprashanth** [comment]
> I do realize things get fuzzy with rcs re-creating pods, but one of the reasons we wanted to add event checking was because of a condition where pods were having issues running and were being killed and restarted. 

I feel like there's a way to address this without event parsing. We recently surfaced container restart'y scenarios in kubectl output (https://github.com/GoogleCloudPlatform/kubernetes/issues/6014).

**Comment by lavalamp** [comment]
@rrati: Scheduler attempted to double-bind that pod. The system is intended to detect and prevent that. This may indicate a subtle scheduler bug or it may indicate that distributed systems don't have perfect informational transfer. Either way, a single occurrence is not a big deal.

As for the second error, the current replication controller is just plain going to make this mistake occasionally; @bprashanth's rewrite should take care of most of these. I'd suggest making a threshold; e.g., maybe 5 of these errors in a run indicate an actual regression.

**Comment by bprashanth** [comment]
IMO events are for human consumption and we'll keep hitting problems like this if we treat them otherwise. They might be the quickest way to get off the ground right now, in which case I'd do as lavalamp recommended and track the counts.

**Comment by rrati** [comment]
@bprashanth This was done before #6014 landed and that may be a better way to determine pod start failures.  If event parsing is as unreliable as it sounds due to acceptable failure events then it makes sense to look at an alternate means to detect issues like these.

@lavalamp Shouldn't the source of the error be investigated if its possible there is a scheduler bug?  It may be low priority and low impact but it doesn't seem like we should just discount it.

The problem I have with a threshold is determining what is reasonable.  If you have 5 errors with 30pods/node on 10 nodes there may be an issue, but 5 errors with 100pods/node on 50 nodes may not be an issue.  Is there some ratio that is sensible based upon something like total pod count maybe?

**Comment by bprashanth** [comment]
Just fyi I'm not saying the test is bad, I get that it was written to serve a purpose :)

**Comment by rrati** [comment]
@bprashanth Yep. :)  #6014 came about from issues running density which sparked the desire to look at events for pod start failures.  If we can get that kind of information w/o looking at events then I think that is a win in multiple ways.

**Comment by lavalamp** [comment]
@rrati the system is self-healing. Distributed systems must be tolerant to different pieces of the system ending up with slightly different states of knowledge. If it happens say more than 10% of the time then maybe we should look into it. Barring that, it's got very low priority because it's just not that big of a deal. The total adverse effect on the system is that scheduler wasted 100ms sending an extra RPC. That's it.

**Comment by lavalamp** [comment]
(Starting too many pods and killing some of them is a much more serious issue and I agree we should fix it, but we're already working on it with @bprashanth's rep. controller rewrite.)

**Comment by quinton-hoole** [comment]
@bprashanth #7236 for fixing automatic e2e cluster log file uploads.

**Comment by wojtek-t** [comment]
>
@rrati the system is self-healing. Distributed systems must be tolerant to different pieces of the system ending up with slightly different states of knowledge. If it happens say more than 10% of the time then maybe we should look into it. Barring that, it's got very low priority because it's just not that big of a deal. The total adverse effect on the system is that scheduler wasted 100ms sending an extra RPC. That's it.

I completely agree - we should definitely log the failures but we should fail the test only of there were more than X failures (at least in case of scheduling failures).

**Comment by rrati** [comment]
@wojtek-t How about we just log all the error/fail messages and add a method to check container status in each pod?  That will better test for the case of container restarting and also give an indication of any possibly odd events.

**Comment by wojtek-t** [comment]
@rrati I'm not sure if I understood correctly. I definitely think that we should log all "fail" events. I even think that we should fail the test if there are too many of those (for some definition of "too many", which we need to decide on).
Do you want to check container statuses for each pod after all of them are started? What do you want to test? I think it's definitely useful to check if they are running. What else you would like to check?

**Comment by rrati** [comment]
@wojtek-t Yes, I think we need to check container statuses for each pod.  We've already seen instances where a pod is marked running but the a container within the pod is continually restarting.  I'd like to verify that all the pods have all the containers running.

I had an e-mail thread at one point about tests we should perform on a loaded system.  I had:
- listing pods, replication controllers, services, minions, and events
- removing a pod
- replication controller restarts a pod
- update a pod
- chaos monkey type test

Would we be differentiating between difference types of "fail" events?  Like a scheduling event is less important than a pod failing to start?

**Comment by wojtek-t** [comment]
>
@wojtek-t Yes, I think we need to check container statuses for each pod. We've already seen instances where a pod is marked running but the a container within the pod is continually restarting. I'd like to verify that all the pods have all the containers running.

SGTM

>
I had an e-mail thread at one point about tests we should perform on a loaded system. I had:
listing pods, replication controllers, services, minions, and events
removing a pod
replication controller restarts a pod
update a pod
chaos monkey type test

Sure - we definitely need more tests. I just didn't get to it yet. Your list SGTM.

>
Would we be differentiating between difference types of "fail" events? Like a scheduling event is less important than a pod failing to start?

I think we we need something like that. Basically if we failed to schedule something because it's already scheduled it doesn't seem to be a serious problem (if there aren't too many of those). Failing to start a pod seems to be a more serious problem.

**Comment by rrati** [comment]
I'm a bit concerned trying to differentiate between different types of failures will lead to a complicated test scenario and introduce flakiness.  I suppose we could look for exact failure type strings (failedScheduling, etc)?  Would those ever change?

**Comment by timothysc** [comment]
Can't we bin on "Reason" and output a histogram of "Reasons"?

**Comment by lavalamp** [comment]
> Can't we bin on "Reason" and output a histogram of "Reasons"?

+1 this is exactly what reason is for. (might want to bin on source+reason, though)

**Comment by timothysc** [comment]
![image](https://cloud.githubusercontent.com/assets/169553/7327967/0b8825e8-ea97-11e4-8864-478de4a58652.png)

Sorry it's Friday ;-)

**Comment by quinton-hoole** [comment]
@wojtek-t  Could you please summarize what the action items from this thread are, and open a separate issue for each?  It seems a lot has been discussed here, but it's not clear to me exactly what concrete action has been decided upon?

**Comment by wojtek-t** [comment]
@quinton-hoole My understanding for what we should do here is:
1. build histogram of events (bin by reason or reason+source)
2. fail the test if there are more than X failures if any category.
3. Check container statuses once all of them are started.
I don't think it makes sense to create a separate issue for those

Another thing mentioned here is to create more sophisticated tests, but I think it's unrelated issue.

**Comment by rrati** [comment]
I'm working on item #3.  Should have  a push ready soon

**Comment by quinton-hoole** [comment]
@wojek-t SGTM

**Comment by wojtek-t** [comment]
@rrati - cool thanks. Please cc me to your PR when it's ready.

**Comment by timothysc** [comment]
latest run with changes : failure on terminated pod. 
http://fpaste.org/216313/02377851/

**Comment by bprashanth** [comment]
@timothysc I've seen the test behave in weird ways, but the ways get weirder with more pods/node. Were you running against a 70+ node cluster?

**Comment by timothysc** [comment]
100 per node is the metric we baseline against.

**Comment by bprashanth** [comment]
I've seen weird situations at that load. Fyi that's outside the 1.0 roadmap, which targets 30-50 per node: https://github.com/GoogleCloudPlatform/kubernetes/blob/b37328a551ba8cdec4dbb3f4f537fce405957c5f/docs/roadmap.md (though we should obviously handle it gracefully).

**Comment by timothysc** [comment]
@bprashanth well aware.. 
FWIW It can be flakey at that level, I've seen it both pass and fail.

**Comment by timothysc** [comment]
@wojtek-t, @rrati - I think this can be closed now.  Thoughts?

**Comment by rrati** [comment]
I think this can be closed as well

**Comment by wojtek-t** [comment]
I agree - closing.
